libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+}
+
+bool initialise_jvm()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    if (!initialise_weak_link_jvm())
+        return false;
+    JavaVMInitArgs vm_args;
+    vm_args.version = JNI_VERSION_1_6;
+    init_jvm_args(&vm_args);
+    
+    JavaVMOption* options = new (nothrow) JavaVMOption[1];
+    options[0].optionString = "-Djava.class.path=/usr/lib/java";
+    
+    vm_args.nOptions = 1;
+    vm_args.options = options;
+    vm_args.ignoreUnrecognized = false;
+    
+    return create_jvm(&vm_args);
+#endif
+    return true;
+}
+
+void finalise_jvm()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    if (s_jvm != nullptr)
+    {
+        s_jvm -> DestroyJavaVM();
+    }
+#endif
+}
+
+void MCJavaDoAttachCurrentThread()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    s_jvm -> AttachCurrentThread((void **)&s_env, nullptr);
+#else
+    s_env = MCJavaAttachCurrentThread();
+#endif
+}
+
+static bool __MCJavaStringFromJString(jstring p_string, MCStringRef& r_string)
+{
+    MCJavaDoAttachCurrentThread();
+    const char *nativeString = s_env -> GetStringUTFChars(p_string, 0);
+    bool t_success = MCStringCreateWithCString(nativeString, r_string);
+    s_env->ReleaseStringUTFChars(p_string, nativeString);
+    return t_success;
+}
+
+static bool __MCJavaStringToJString(MCStringRef p_string, jstring& r_string)
+{
+    MCJavaDoAttachCurrentThread();
+    
+    MCAutoStringRefAsCString t_string_cstring;
+    t_string_cstring . Lock(p_string);
+    
+    jstring t_result;
+    t_result = s_env->NewStringUTF(*t_string_cstring);
+    
+    if (t_result != nullptr)
+    {
+        r_string = t_result;
+        return true;
+    }
+    
+    return false;
+}
+
+static bool __MCJavaDataFromJByteArray(jbyteArray p_byte_array, MCDataRef& r_data)
+{
+    MCJavaDoAttachCurrentThread();
+    
+    jbyte *t_bytes = nullptr;
+    uint32_t t_length = 0;
+    
+    if (p_byte_array != nullptr)
+        t_bytes = s_env -> GetByteArrayElements(p_byte_array, nullptr);
+    
+    if (t_bytes == nullptr)
+        return false;
+
+    t_length = s_env -> GetArrayLength(p_byte_array);
+    bool t_success = MCDataCreateWithBytes((const byte_t *)t_bytes, t_length, r_data);
+    s_env -> ReleaseByteArrayElements(p_byte_array, t_bytes, 0);
+
+    return t_success;
+}
+
+static bool __MCJavaDataToJByteArray(MCDataRef p_data, jbyteArray& r_byte_array)
+{
+    MCJavaDoAttachCurrentThread();
+    
+    if (p_data == nullptr || MCDataIsEmpty(p_data))
+    {
+        r_byte_array = nullptr;
+        return true;
+    }
+    
+    uindex_t t_length = MCDataGetLength(p_data);
+    jbyteArray t_bytes = s_env -> NewByteArray(t_length);
+    
+    if (t_bytes == nullptr)
+        return false;
+    
+    const byte_t* t_data = MCDataGetBytePtr(p_data);
+    s_env -> SetByteArrayRegion(t_bytes, 0, t_length,
+                                reinterpret_cast<const jbyte*>(t_data));
```

Comment:
To avoid a strict aliasing violation, I suggest writing this:

```c++
auto t_data = reinterpret_cast<const jbyte*>(MCDataGetBytePtr(p_data));
s_env->SetByteArrayRegion(t_bytes, 0, t_length, t_data);
```

I think that will avoid any possibility of the compiler deciding that you are accessing `t_data` via two distinct pointer types.